File: task_5.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> continued_fraction(int a, int b) {
    vector<int> coefficients;
    while (b != 0) {
        int q = a / b;
        coefficients.push_back(q);
        int r = a % b;
        a = b;
        b = r;
    }
    return coefficients;
}

int EVK(int a, int b) {
    while (b != 0) {
        int r = b;
        b = a % b;
        a = r;
    }
    
    return a;
}

void CompactSolution(int a, int a0, int b, int b0){
    int k = ( (a / a0) < (b / b0) ) ? (a / a0) : (b / b0);

    cout << "Компактное общее решение: a = " << a + (a0 * k) << " + " << a0 << " * k"
        << ", b = " << b + (-b0 * k) << " + " << b0 << " * k"
        << ", где k - целое число." << endl;
}

void solve_diophantine(int a, int b, int c) {
    vector<int> cf = continued_fraction(a, b);
    int x = 0, y = 1;
    int x_prev = 1, y_prev = 0;
    for (int q : cf) {
        int x_temp = x;
        x = x_prev - q * x;
        x_prev = x_temp;
        
        int y_temp = y;
        y = y_prev - q * y;
        y_prev = y_temp;
    }

    int gcd = EVK(a, b);
    if (c % gcd != 0) {
        cout << "Нет решений." << endl;
        return;
    }
    
    int k = c / gcd;
    cout << "Одно решение: a = " << x_prev * k << ", b = " << y_prev * k << endl;
    cout << "Общее решение: a = " << x_prev * k << " + " << b/gcd << " * t," 
         << " b = " << y_prev * k << " - " << a/gcd << " * t" 
         << ", где t - целое число." << endl;

    CompactSolution(x_prev * k, b/gcd, y_prev * k, a/gcd);
}

int main() {
    cout << "1256a + 847b = 119" << endl;
    solve_diophantine(1256, 847, 119);

    return 0;
}

File: pyhton_solutions/task_5.py
from typing import List


def continued_fraction(a: int, b: int) -> List[int]:
    coefficients = []
    while b != 0:
        q = a // b
        coefficients.append(q)
        r = a % b
        a, b = b, r
    return coefficients


def gcd(a: int, b: int) -> int:
    while b != 0:
        r = b
        b = a % b
        a = r
    return a


def compact_solution(a: int, a0: int, b: int, b0: int):
    k = min(a // a0, b // b0) if (a0 != 0 and b0 != 0) else 0

    print(f"Одно компактное решение: a = {a + a0 * k}, b = {b - b0 * k}")
    print(f"Компактное общее решение: a = {a + a0 * k} + {a0} * k, "
          f"b = {b - b0 * k} + {b0} * k, где k - целое число.")


def solve_diophantine(a: int, b: int, c: int):
    cf = continued_fraction(a, b)
    x, y = 0, 1
    x_prev, y_prev = 1, 0

    for q in cf:
        x, x_prev = x_prev - q * x, x
        y, y_prev = y_prev - q * y, y

    d = gcd(a, b)
    if c % d != 0:
        print("Нет решений.")
        return

    k = c // d
    a0 = b // d
    b0 = a // d

    print(f"Одно решение: x = {x_prev * k}, y = {y_prev * k}")
    print(f"Общее решение: x = {x_prev * k} + {a0} * k, "
          f"y = {y_prev * k} - {b0} * k, где k - целое число.\n")

    compact_solution(x_prev * k, a0, y_prev * k, b0)


if __name__ == "__main__":
    print("1256x + 847y = 119\n")
    solve_diophantine(1256, 847, 119)